#Task 3

The file “Parts.csv” contains descriptions of some fictitious parts. Your goal is to
find 5 alternative parts to each provided fictitious part in the dataset based on their
similarity. First provide descriptive analysis of the data and highlight 2-3 findings
and difficulties of the data that we provided and describe how you would handle this.
Continue to implement a solution that is finding the similar fictitious parts based on
the column “DESCRIPTION”. Please give details of your solution and why you choose it.
Once you finished your implementation of your solution, please think about how you
would integrate your code into the chatbot from task 1.

##Goal = Aanalyze Parts.csv dataset of fictitious parts and find alternative parts that are similar to each given part using LLM.

1. Descriptive Analysis: Examine the data to understand its structure, identify patterns, and note any challenges or limitations.

2. Similarity Matching: Use GenAI models to acheive this.

3. Integration: Suggest how this solution could be integrated into a chatbot for real-time use.

### This is useful in scenarios like inventory management, product recommendation, or customer support, where a user might ask, "What are alternatives to part A1?" The chatbot would analyze the data and provide similar parts based on their descriptions using text generation Model.

In [3]:
from google.colab import drive
drive.mount('/content/drive')
import os

Mounted at /content/drive


In [1]:
import platform
# checking the system info
print("Processor:", platform.processor())
print("Machine:", platform.machine())
print("Platform:", platform.platform())
print("System:", platform.system())
print("Version:", platform.version())

Processor: x86_64
Machine: x86_64
Platform: Linux-6.1.123+-x86_64-with-glibc2.35
System: Linux
Version: #1 SMP PREEMPT_DYNAMIC Sun Mar 30 16:01:29 UTC 2025


In [2]:
import torch
print(torch.cuda.is_available()) #using GPU

True


In [4]:
## Libraries Required
!pip install langchain-huggingface
## For API Calls
!pip install huggingface_hub
!pip install transformers
!pip install accelerate
!pip install  bitsandbytes
!pip install langchain

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 87.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 93.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 65.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 13.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 9.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 5.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 108.1 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstalling nvidia-nvjitlink-cu12-12.5.82:
      Successfully uninstalled nvidia-nvjitli

In [ ]:
## Environment secret keys
from google.colab import userdata
sec_key=userdata.get("HF_TOKEN")
print(sec_key)

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


In [5]:
# Importing required generic libraries
import warnings
warnings.filterwarnings("ignore")

import re
import pandas as pd
import numpy as np
from tabulate import tabulate

from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity
import torch
#PyTorch is used by Hugging Face Transformers under the hood for model loading and inference (like with Mistral or LLaMA models).
from langchain_huggingface import HuggingFaceEndpoint
from langchain_huggingface import HuggingFacePipeline
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig, pipeline
from transformers import BitsAndBytesConfig

##Initializing Sentence Transformer Model

In [6]:
global similarity_model,llm_tokenizer, llm_model, llm_pipeline
SIMILARITY_MODEL_NAME = 'all-MiniLM-L6-v2'
similarity_model = SentenceTransformer(SIMILARITY_MODEL_NAME)

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

## HuggingFaceEndpoint - Access HuggingFace Models with API
There are two ways to use this class. You can specify the model with the repo_id parameter. Those endpoints use the serverless API, which is particularly beneficial to people using pro accounts or enterprise hub. Still, regular users can already have access to a fair amount of request by connecting with their HF token in the environment where they are executing the code.


In [7]:
from google.colab import userdata
sec_key=userdata.get("HUGGINGFACEHUB_API_TOKEN")
print(sec_key)

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


In [ ]:
os.environ["HUGGINGFACEHUB_API_TOKEN"]=sec_key

In [ ]:
repo_id="mistralai/Mistral-7B-Instruct-v0.2"
llm=HuggingFaceEndpoint(repo_id=repo_id,max_length=128,temperature=0.7,token=sec_key)

                    max_length was transferred to model_kwargs.
                    Please make sure that max_length is what you intended.
                    token was transferred to model_kwargs.
                    Please make sure that token is what you intended.


In [ ]:
llm.invoke("What is machine learning")

TypeError: InferenceClient.text_generation() got an unexpected keyword argument 'max_length'

## HuggingFacePipeline - class from LangChain that allows us to use Hugging Face models (especially text generation pipelines) as LLMs in LangChain workflows.


Among transformers, the Pipeline is the most versatile tool in the Hugging Face toolbox. LangChain being designed primarily to address RAG and Agent use cases, the scope of the pipeline here is reduced to the following text-centric tasks: “text-generation", “text2text-generation", “summarization”, “translation”.
Models can be loaded directly with the from_model_id method


## Trying gpt2 model considering resources. But will use Mistral for better Accuracy as it is the state of the art model currently.

In [ ]:
model_id="gpt2"
llm_model=AutoModelForCausalLM.from_pretrained(model_id)
llm_tokenizer=AutoTokenizer.from_pretrained(model_id)

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

In [ ]:
llm_pipeline = pipeline("text-generation",model=llm_model,tokenizer=llm_tokenizer,max_new_tokens=256)
hf=HuggingFacePipeline(pipeline=llm_pipeline)

Device set to use cpu


In [ ]:
hf

HuggingFacePipeline(pipeline=<transformers.pipelines.text_generation.TextGenerationPipeline object at 0x7b4876ffda10>, model_id='gpt2')

In [ ]:
hf.invoke("What is machine learning")

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


'What is machine learning?\n\nMachine learning is a computational process that enables us to create new and interesting products, services, and technologies. Machine learning is a form of object-oriented programming using techniques that allow us to manipulate objects and their data without needing to write code.\n\nMachine learning is also used to teach us skills that we can use to perform other useful tasks.\n\nMachine learning is also used to teach us new and interesting products, services, and technologies. Machine learning is also used to teach us skills that we can use to perform other useful tasks. Machine learning is used to teach us the power of data, in many cases, and to build systems to build better, more powerful systems.\n\nMachine learning is a form of object-oriented programming that enables us to create new and interesting products, services, and technologies. Machine learning is also used to teach us new and interesting products, services, and technologies. Machine le

## Mistral Model Loading - Considering 4bit Quatization.

In [8]:
model_id = "mistralai/Mistral-7B-Instruct-v0.1"

# Quantization config (4-bit)
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",  # best quant type for performance
)

# Load tokenizer
llm_tokenizer = AutoTokenizer.from_pretrained(model_id, use_fast=True)

# Load model with quantization
llm_model = AutoModelForCausalLM.from_pretrained(
    model_id,
    quantization_config=bnb_config,
    device_map="auto",  # Auto-assigns layers to GPU if available
    trust_remote_code=True
)

#use below block if GPU is not accessible
'''
llm_model = AutoModelForCausalLM.from_pretrained(
    model_id,
    torch_dtype=torch.float16,  # or float32 if needed
    device_map="auto",
)
'''

tokenizer_config.json:   0%|          | 0.00/2.10k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/493k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.80M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/414 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/25.1k [00:00<?, ?B/s]

Fetching 2 files:   0%|          | 0/2 [00:00<?, ?it/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/9.94G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/4.54G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

'\nllm_model = AutoModelForCausalLM.from_pretrained(\n    model_id,\n    torch_dtype=torch.float16,  # or float32 if needed\n    device_map="auto",\n)\n'

In [25]:
def generate_llm_response_mistral(prompt):
    inputs = llm_tokenizer(prompt, return_tensors="pt").to(llm_model.device)
    outputs = llm_model.generate(**inputs, max_new_tokens=256)
    return llm_tokenizer.decode(outputs[0], skip_special_tokens=True)

In [10]:
# Example usage
context="A1 Information about Part ID A1: DESCRIPTION: Indicator Red Fast Movement 1.6A 250V Holder Plastic 5 X 20mm Ceramic Box CCC/PSE/VDE/cULus Electric Indicator, Very Fast Blow, 1.6A, 250VAC, 1500A (IR), Inline/holder, 5x20mm. Attribut1: Fast. Application: Primary Protection In Equipment. Characteristic: VERY FAST. Height: 20mm. Length in mm: 5.2mm. Rating: 1.6A. Material: Ceramic. Size: 5 X 20mm. Joule-integral-Nom (J): 0.755J. LC Risk: Low. Maximum AC Voltage Rating: 250V. Maximum Power Dissipation: 4W. Mounting: Holder. Mounting Feature: INLINE/HOLDER. Number of Terminals: 2.0. Operating Temperature-Max (Cel): 125Cel. Operating Temperature-Min (Cel): -55Cel. Physical Dimension: 5.2mm  x  20mm. Pre-arcing time-Min (ms): 3ms. Product Diameter: 5.2mm. Product Length: 20mm. Rated Breaking Capacity (A): 1500A. Rated Current (A): 1.6A. Rated Voltage (V): 250V. Rated Voltage(AC) (V): 250V."
user_query = "Tell me about A1"
prompt = f"{context}\n\nUser: {user_query}\nAI:"
response = generate_llm_response(prompt)
print(response)

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


A1 Information about Part ID A1: DESCRIPTION: Indicator Red Fast Movement 1.6A 250V Holder Plastic 5 X 20mm Ceramic Box CCC/PSE/VDE/cULus Electric Indicator, Very Fast Blow, 1.6A, 250VAC, 1500A (IR), Inline/holder, 5x20mm. Attribut1: Fast. Application: Primary Protection In Equipment. Characteristic: VERY FAST. Height: 20mm. Length in mm: 5.2mm. Rating: 1.6A. Material: Ceramic. Size: 5 X 20mm. Joule-integral-Nom (J): 0.755J. LC Risk: Low. Maximum AC Voltage Rating: 250V. Maximum Power Dissipation: 4W. Mounting: Holder. Mounting Feature: INLINE/HOLDER. Number of Terminals: 2.0. Operating Temperature-Max (Cel): 125Cel. Operating Temperature-Min (Cel): -55Cel. Physical Dimension: 5.2mm  x  20mm. Pre-arcing time-Min (ms): 3ms. Product Diameter: 5.2mm. Product Length: 20mm. Rated Breaking Capacity (A): 1500A. Rated Current (A): 1.6A. Rated Voltage (V): 250V. Rated Voltage(AC) (V): 250V.

User: Tell me about A1
AI: A1 is an electric indicator that is designed to provide fast protection in eq

In [11]:
# Load the datasets
df_original = pd.read_csv('/content/drive/MyDrive/BMW_csv_files/Parts.csv', sep=';')

In [12]:
# Adjust display options for single-line output
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)

# confirm the data loaded: Print the dataframe in a single line
print(df_original.head().to_string(index=False))

ID                                                                                                                                                                     DESCRIPTION Attribut1 Additional Feature                     Application Characteristic Temp Height Length in mm Rating Material     Size Code Joule-integral-Nom (J) LC Risk Maximum AC Voltage Rating Maximum DC Voltage Rating Maximum Power Dissipation Mounting Mounting Feature  Number of Terminals Operating Temperature-Max (Cel) Operating Temperature-Min (Cel) Physical Dimension Pre-arcing time-Min (ms) Product Diameter Product Length Rated Breaking Capacity (A) Rated Current (A) Rated Voltage (V) Rated Voltage(AC) (V) Rated Voltage(DC) (V)
A1 Indicator Red Fast Movement 1.6A 250V Holder Plastic 5 X 20mm Ceramic Box CCC/PSE/VDE/cULus Electric Indicator, Very Fast Blow, 1.6A, 250VAC, 1500A (IR), Inline/holder, 5x20mm      Fast                NaN Primary Protection In Equipment      VERY FAST  NaN   20mm        5.2mm   1.6A

In [13]:
num_rows, num_columns = df_original.shape
print(f"There are {num_rows} rows (records) and {num_columns} columns (fields) in the dataset.")

There are 998 rows (records) and 32 columns (fields) in the dataset.


## 1. Descriptive Analysis of the Data.
The "DESCRIPTION" column is the primary focus. Descriptions use natural language (e.g., "Indicator Red Fast Movement" vs. "Non Resettable Indicators Electric Indicator"), which may lead to parsing challenges.

In [ ]:
df_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 998 entries, 0 to 997
Data columns (total 32 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   ID                               998 non-null    object 
 1   DESCRIPTION                      663 non-null    object 
 2   Attribut1                        770 non-null    object 
 3   Additional Feature               324 non-null    object 
 4   Application                      821 non-null    object 
 5   Characteristic                   672 non-null    object 
 6   Temp                             417 non-null    object 
 7   Height                           705 non-null    object 
 8   Length in mm                     705 non-null    object 
 9   Rating                           834 non-null    object 
 10  Material                         771 non-null    object 
 11  Size                             880 non-null    object 
 12  Code                  

In [ ]:
df_original.describe(include='all').T

count unique                                                top freq mean  std  min  25%  50%  75%  max
ID                                 998    998                                               A998    1  NaN  NaN  NaN  NaN  NaN  NaN  NaN
DESCRIPTION                        663    582  Indicator Chip Very Fast Movement 3A 125V SMD ...    4  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Attribut1                          770      5                                          Very Fast  313  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Additional Feature                 324     39           RATED BREAKING CAPACITY AT 125 VDC: 50 A  129  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Application                        821     28                                      Motor Circuit  293  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Characteristic                     672      7                                          VERY FAST  231  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Temp                               417     16                                             2.69mm  187  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Height                             705     33                                               20mm  189  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Length in mm                       705     30                                              5.2mm  187  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Rating                             834     70                                                 4A   50  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Material                           771      8                                            Ceramic  420  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Size                               880     39                                           5 X 20mm  187  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Code                               555      4                                                 e3  207  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Joule-integral-Nom (J)             663    374                                              0.53J    5  NaN  NaN  NaN  NaN  NaN  NaN  NaN
LC Risk                            786      3                                                Low  566  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Maximum AC Voltage Rating          734      9                                               250V  338  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Maximum DC Voltage Rating          448     11                                               125V  219  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Maximum Power Dissipation          253    103                                          1.6(Typ)W   38  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Mounting                           774      5                                      Surface Mount  467  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Mounting Feature                   679      3                                      SURFACE MOUNT  435  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Number of Terminals              758.0    NaN                                                NaN  NaN  2.0  0.0  2.0  2.0  2.0  2.0  2.0
Operating Temperature-Max (Cel)    707      5                                             125Cel  529  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Operating Temperature-Min (Cel)    702      2                                             -55Cel  677  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Physical Dimension                 668     38                        6.1mm  x  2.69mm  x  2.69mm  182  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Pre-arcing time-Min (ms)           108     18                                               10ms   49  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Product Diameter                   272     15                                              5.2mm  176  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Product Length                     760     38                                              6.1mm  202  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Rated Breaking Capacity (A)        655     19                                                50A  351  NaN  NaN  NaN  NaN  NaN  NaN  NaN
Rated Current (A)                  673     52               

In [ ]:
df_original['DESCRIPTION'].head().to_list()

['Indicator Red Fast Movement 1.6A 250V Holder Plastic 5 X 20mm Ceramic Box CCC/PSE/VDE/cULus Electric Indicator, Very Fast Blow, 1.6A, 250VAC, 1500A (IR), Inline/holder, 5x20mm',
 'Non Resettable Indicators Electric Indicator, Very Fast Blow, 6.3A, 250VAC, 1500A (IR), Inline/holder, 5x20mm',
 'Indicator Red Fast Movement 8A 250V Holder Plastic 5 X 20mm Ceramic Box KC/PSE/VDE/cULus Electric Indicator, Very Fast Blow, 8A, 250VAC, 1500A (IR), Inline/holder, 5x20mm',
 'Non Resettable Indicators Electric Indicator, Very Fast Blow, 10A, 250VAC, 1500A (IR), Inline/holder, 5x20mm',
 'Indicator Red Fast Movement 12.5A 250V Holder Plastic 5 X 20mm Ceramic Box PSE/cULus Electric Indicator, Very Fast Blow, 12.5A, 250VAC, 500A (IR), Inline/holder, 5x20mm']

In [ ]:
# Check duplicates in the data frame
print(df_original[df_original.duplicated()])
print("There are",df_original.duplicated().sum(), "duplicates")

Empty DataFrame
Columns: [ID, DESCRIPTION, Attribut1, Additional Feature, Application, Characteristic, Temp, Height, Length in mm, Rating, Material, Size, Code, Joule-integral-Nom (J), LC Risk, Maximum AC Voltage Rating, Maximum DC Voltage Rating, Maximum Power Dissipation, Mounting, Mounting Feature, Number of Terminals, Operating Temperature-Max (Cel), Operating Temperature-Min (Cel), Physical Dimension, Pre-arcing time-Min (ms), Product Diameter, Product Length, Rated Breaking Capacity (A), Rated Current (A), Rated Voltage (V), Rated Voltage(AC) (V), Rated Voltage(DC) (V)]
Index: []
There are 0 duplicates


###No Duplicates, Great!

In [ ]:
# Checking for null/missing values in the data
missing_values = df_original.isnull().sum()
if missing_values.sum() == 0:
    print("\nThere are NO missing values in the dataset.\n")
else:
    print("\nThere are missing values in the dataset which need treatment.\n")
    print("\nMissing values per column:\n")
    print(missing_values)


There are missing values in the dataset which need treatment.


Missing values per column:

ID                                   0
DESCRIPTION                        335
Attribut1                          228
Additional Feature                 674
Application                        177
Characteristic                     326
Temp                               581
Height                             293
Length in mm                       293
Rating                             164
Material                           227
Size                               118
Code                               443
Joule-integral-Nom (J)             335
LC Risk                            212
Maximum AC Voltage Rating          264
Maximum DC Voltage Rating          550
Maximum Power Dissipation          745
Mounting                           224
Mounting Feature                   319
Number of Terminals                240
Operating Temperature-Max (Cel)    291
Operating Temperature-Min (Cel)    296
Physical D

#2. Observations and Findings

Structure: The dataset contains 998 entries with columns such as "ID," "DESCRIPTION," "Additional Feature," "Application," "Characteristic," "Rated Current (A)," "Rated Voltage (AC) (V)," "Rated Voltage (DC) (V)," "Material," "Size," and more. The "DESCRIPTION" column provides detailed specifications (e.g., "Indicator Red Fast Movement 1.6A 250V Holder Plastic 5 X 20mm Ceramic Box CCC/PSE/VDE/cULus Electric Indicator").

Variability: Parts vary by current (0.002A to 30A), voltage (72V to 500V), blow type (Very Fast Blow, Time Lag Blow, Slow Blow), material (Ceramic, Glass), and mounting type (Inline/Holder, Surface Mount, Through Hole).

Data Quality: Some entries (e.g., A8, A18) have incomplete "DESCRIPTION" fields, while others (e.g., A63-A79) are accessory holders with no current or voltage ratings.

##Data cleansing

In [14]:
parts_df= df_original.copy(deep=True)

In [15]:
parts_df.replace('', np.nan, inplace=True) # Replace empty strings with NaN

In [16]:
#function to preprocess text
def preprocess_text_for_similarity(text):
    """Cleans description text for embedding."""
    text = str(text).lower()
    text = re.sub(r'[^\w\s\-\.\/x]', '', text) # Keep numbers, letters, spaces, -, ., /, x
    text = re.sub(r'\s+', ' ', text).strip() # Replace multiple spaces with single
    return text

In [17]:
# Precompute cleaned descriptions and embeddings for faster similarity search
parts_df.dropna(subset=['DESCRIPTION'], inplace=True)
parts_df['CLEANED_DESCRIPTION'] = parts_df['DESCRIPTION'].apply(preprocess_text_for_similarity)

In [18]:
parts_df[["DESCRIPTION","CLEANED_DESCRIPTION"]].head()

DESCRIPTION                                CLEANED_DESCRIPTION
0  Indicator Red Fast Movement 1.6A 250V Holder P...  indicator red fast movement 1.6a 250v holder p...
1  Non Resettable Indicators Electric Indicator, ...  non resettable indicators electric indicator v...
2  Indicator Red Fast Movement 8A 250V Holder Pla...  indicator red fast movement 8a 250v holder pla...
3  Non Resettable Indicators Electric Indicator, ...  non resettable indicators electric indicator v...
4  Indicator Red Fast Movement 12.5A 250V Holder ...  indicator red fast movement 12.5a 250v holder ...

In [19]:
parts_df['EMBEDDING'] = list(similarity_model.encode(parts_df['CLEANED_DESCRIPTION'].tolist(), show_progress_bar=True))
#CLEANED_DESCRIPTION series is converted to list and encoded into embeddings(384 features).

Batches:   0%|          | 0/21 [00:00<?, ?it/s]

In [20]:
len(parts_df['EMBEDDING'][0])

384

In [21]:
num_rows, num_columns = parts_df.shape
print(f"There are {num_rows} rows (records) and {num_columns} columns (fields) in the dataset.")

There are 663 rows (records) and 34 columns (fields) in the dataset.


In [22]:
user_input="Tell me about A1"
#Try to find a specific part ID
part_id_match = re.search(r'\b([aA]\d+)\b', user_input)
print(part_id_match)

<re.Match object; span=(14, 16), match='A1'>


In [23]:
def generate_context_for_part(part_id):
    """Generates a textual context for a given part ID."""
    if parts_df is None:
        return "Data not loaded. Please check the data file."

    part_info = parts_df[parts_df['ID'].str.lower() == part_id.lower()]

    if part_info.empty:
        return None # Indicate part not found

    part_data = part_info.iloc[0].dropna().to_dict()
    #part_data : {'ID': 'A10', 'DESCRIPTION': 'Non Resettable Indicators Electric Indicator, Time Lag Blow, 2A, 250VAC, 300VDC, 1500A (IR), Inline/holder, 5x20mm', 'Additional Feature': 'RATED BREAKING CAPACITY AT 300 VDC: 1500 A'}
    context = f"Information about Part ID {part_id}:\n"
    for key, value in part_data.items():
        if key not in ['ID', 'CLEANED_DESCRIPTION', 'EMBEDDING']: # Exclude internal columns
            context += f"{key}: {value}. "

    return context.strip()

In [24]:
if part_id_match:
  part_id = part_id_match.group(1).upper()
  print(part_id)
  part_context = generate_context_for_part(part_id)
  if part_context is None:
    context = f"Part with ID '{part_id}' not found in the database."
  else:
    context = part_context

print(context)

A1
Information about Part ID A1:
DESCRIPTION: Indicator Red Fast Movement 1.6A 250V Holder Plastic 5 X 20mm Ceramic Box CCC/PSE/VDE/cULus Electric Indicator, Very Fast Blow, 1.6A, 250VAC, 1500A (IR), Inline/holder, 5x20mm. Attribut1: Fast. Application: Primary Protection In Equipment. Characteristic: VERY FAST. Height: 20mm. Length in mm: 5.2mm. Rating: 1.6A. Material: Ceramic. Size: 5 X 20mm. Joule-integral-Nom (J): 0.755J. LC Risk: Low. Maximum AC Voltage Rating: 250V. Maximum Power Dissipation: 4W. Mounting: Holder. Mounting Feature: INLINE/HOLDER. Number of Terminals: 2.0. Operating Temperature-Max (Cel): 125Cel. Operating Temperature-Min (Cel): -55Cel. Physical Dimension: 5.2mm  x  20mm. Pre-arcing time-Min (ms): 3ms. Product Diameter: 5.2mm. Product Length: 20mm. Rated Breaking Capacity (A): 1500A. Rated Current (A): 1.6A. Rated Voltage (V): 250V. Rated Voltage(AC) (V): 250V.


In [39]:
def generate_llm_response_test(user_query, context=""):
    prompt = f"Context:\n{context}\n\nQuestion: {user_query}\nAnswer:"
    output = hf(prompt)
    return output.strip()

In [ ]:
# Generate response
response = generate_llm_response_test(user_input, context)
print(response)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


Context:
Information about Part ID A1:
DESCRIPTION: Indicator Red Fast Movement 1.6A 250V Holder Plastic 5 X 20mm Ceramic Box CCC/PSE/VDE/cULus Electric Indicator, Very Fast Blow, 1.6A, 250VAC, 1500A (IR), Inline/holder, 5x20mm. Attribut1: Fast. Application: Primary Protection In Equipment. Characteristic: VERY FAST. Height: 20mm. Length in mm: 5.2mm. Rating: 1.6A. Material: Ceramic. Size: 5 X 20mm. Joule-integral-Nom (J): 0.755J. LC Risk: Low. Maximum AC Voltage Rating: 250V. Maximum Power Dissipation: 4W. Mounting: Holder. Mounting Feature: INLINE/HOLDER. Number of Terminals: 2.0. Operating Temperature-Max (Cel): 125Cel. Operating Temperature-Min (Cel): -55Cel. Physical Dimension: 5.2mm  x  20mm. Pre-arcing time-Min (ms): 3ms. Product Diameter: 5.2mm. Product Length: 20mm. Rated Breaking Capacity (A): 1500A. Rated Current (A): 1.6A. Rated Voltage (V): 250V. Rated Voltage(AC) (V): 250V.

Question: Tell me about A1
Answer: A1 is an A2 holder. The A2 is a 4 x 16 x 24 mm (or 0.55" by 0.5

Why GPT-2 is doing this?

1. GPT-2 is not instruction-tuned, so it doesn't "understand"  structured Q&A very well.
2. Prone to repetition
3. Limited in capability on complex or structured inputs

| Model Name                           | Type         | Size | Instruction-Tuned | Speed       | Accuracy     | Usage Notes                                          |
| ------------------------------------ | ------------ | ---- | ----------------- | ----------- | ------------ | ---------------------------------------------------- |
| `gpt2`                               | Decoder-only | 124M | ❌ No              | ⚡ Very Fast | ❌ Low        | Not suitable for Q\&A or chat; general text gen only |
| `tiiuae/falcon-7b-instruct`          | Decoder-only | 7B   | ✅ Yes             | ⚡ Moderate  | ✅ Good       | Great open-source instruction model                  |
| `mistralai/Mistral-7B-Instruct-v0.1` | Decoder-only | 7B   | ✅ Yes             | ⚡⚡ Fast     | ✅✅ Very Good | Among the best open-source models                    |
| `databricks/dolly-v2-7b`             | Decoder-only | 7B   | ✅ Yes             | ⚡ Moderate  | ⚠️ Outdated  | Trained on open Q\&A data; less capable than Mistral |


## Mistral Output

In [30]:
llm_pipeline = pipeline("text-generation", model=llm_model, tokenizer=llm_tokenizer, max_new_tokens=256, device_map="auto")
print("Mistral LLM initialized.")

Device set to use cuda:0


Mistral LLM initialized.


In [31]:
def generate_llm_response(user_query, context=""):
    # Construct the prompt for Mistral
    system_prompt = "You are a helpful assistant for a parts catalog. Answer questions accurately based *only* on the provided context. If the information is not in the context, state that you don't know or cannot find the information."

    if context:
        full_prompt = f"<s>[INST] {system_prompt}\n\nContext:\n{context}\n\nUser Question: {user_query} [/INST]"
    else:
        full_prompt = f"<s>[INST] {system_prompt}\n\nUser Question: {user_query} [/INST]"

    response = llm_pipeline(full_prompt)[0]['generated_text']
    # Extract only the model's answer part if the pipeline returns the full prompt + answer
    if "[/INST]" in response:
      response = response.split("[/INST]")[-1].strip()
      return response

In [32]:
# Generate response
response = generate_llm_response(user_input, context)
print(response)

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


A1 is an electric indicator that is used for primary protection in equipment. It is designed to provide fast detection of electrical faults and has a very fast blow characteristic. The indicator has a ceramic body, measures 5 x 20mm in size and is rated at 1.6A, 250VAC, 1500A, and 4W. It has two terminals and operates within a temperature range of -55°C to 125°C. The pre-arcing time is 3ms, and it has a rated breaking capacity of 1500A and a rated current of 1.6A.


In [33]:
# Adjust display options for single-line output
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)

# Generate response
response = generate_llm_response(user_input, context)
print(response)


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


A1 is an Electric Indicator with the following specifications:

* Description: Indicator Red Fast Movement 1.6A 250V Holder Plastic 5 X 20mm Ceramic Box CCC/PSE/VDE/cULus Electric Indicator, Very Fast Blow, 1.6A, 250VAC, 1500A (IR), Inline/holder, 5x20mm.
* Height: 20mm.
* Length in mm: 5.2mm.
* Rating: 1.6A.
* Material: Ceramic.
* Size: 5 X 20mm.
* Joule-integral-Nom (J): 0.755J.
* LC Risk: Low.
* Maximum AC Voltage Rating: 250V.
* Maximum Power Dissipation: 4W.
* Mounting: Holder.
* Mounting Feature: INLINE/HOLDER.
* Number of Terminals: 2.
* Operating Temperature-Max (C


In [34]:
user_input="Find 5 alternative parts for part A2 "
#Try to find a specific part ID
part_id_match = re.search(r'\b([aA]\d+)\b', user_input)
print(part_id_match)

<re.Match object; span=(34, 36), match='A2'>


In [35]:
if part_id_match:
  part_id = part_id_match.group(1).upper()
  print(part_id)
  part_context = generate_context_for_part(part_id)
  if part_context is None:
    context = f"Part with ID '{part_id}' not found in the database."
  else:
    context = part_context

print(context)

A2
Information about Part ID A2:
DESCRIPTION: Non Resettable Indicators Electric Indicator, Very Fast Blow, 6.3A, 250VAC, 1500A (IR), Inline/holder, 5x20mm. Application: Primary Protection In Equipment. Characteristic: VERY FAST. Height: 20mm. Length in mm: 5.2mm. Material: Ceramic. Size: 5 X 20mm. Code: e2. Joule-integral-Nom (J): 36.7J. LC Risk: Low. Mounting: Holder. Mounting Feature: INLINE/HOLDER. Number of Terminals: 2.0. Operating Temperature-Max (Cel): 125Cel. Operating Temperature-Min (Cel): -55Cel. Physical Dimension: 5.2mm  x  20mm. Pre-arcing time-Min (ms): 3ms. Product Diameter: 5.2mm. Product Length: 20mm. Rated Breaking Capacity (A): 1500A. Rated Current (A): 6.3A. Rated Voltage (V): 250V. Rated Voltage(AC) (V): 250V.


In [38]:
# Adjust display options for single-line output
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)

# Generate response
response = generate_llm_response(user_input, context)
print(response)

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


Based on the information provided, here are 5 alternative parts to Part A2:

1. Part ID A1 - Non Resettable Indicators Electric Indicator, Fast Blow, 6.3A, 250VAC, 1500A (IR), Inline/holder, 5x20mm.
2. Part ID A3 - Non Resettable Indicators Electric Indicator, Fast Blow, 6.3A, 250VAC, 1500A (IR), Inline/holder, 5x20mm.
3. Part ID A4 - Non Resettable Indicators Electric Indicator, Fast Blow, 6.3A, 250VAC, 1500A (IR), Inline/holder, 5x20mm.
4. Part ID A5 - Non Resettable Indicators Electric Indicator, Fast Blow, 6.3A, 250VAC, 1500A (IR), Inline/holder, 5x20mm.
5. Part ID A6 - Non Resettable Indicators


##Justification for Using the Mistral Model
The Mistral model, a high-performing open-source LLM developed by Mistral AI, is a suitable choice for this task due to the following reasons:

1. Efficiency and Performance: Mistral-7B is designed to deliver high performance with relatively low computational requirements compared to larger models like GPT-4. This efficiency is critical for processing the "Parts.csv" dataset and generating embeddings or responses quickly, especially in a resource-constrained environment like Google Colab with GPU support (as verified in the notebook).

2. Text Understanding and Generation: Mistral excels in natural language understanding and generation, making it ideal for analyzing part descriptions and identifying semantic similarities. Its ability to capture nuanced relationships in text ensures accurate similarity matching for the "DESCRIPTION" column, which is essential for recommending alternative parts.

3. Hugging Face Integration: The notebook leverages the Hugging Face ecosystem, which supports Mistral models through transformers and langchain-huggingface. This seamless integration simplifies model loading, fine-tuning, and deployment, as seen with the use of HuggingFacePipeline and AutoModelForCausalLM. Mistral's compatibility with these tools streamlines the development process.

4. Quantization Support: The use of BitsAndBytesConfig in the notebook suggests quantization (e.g., 4-bit or 8-bit) to reduce memory usage while maintaining performance. Mistral models are well-suited for quantization, allowing efficient deployment on hardware with limited resources, which aligns with the notebook’s setup.

5. Open-Source Accessibility: As an open-source model, Mistral is freely accessible via Hugging Face (using the provided HF_TOKEN), making it cost-effective and customizable compared to proprietary models. This is particularly advantageous for a project requiring experimentation and integration into a chatbot without incurring high API costs.

6. Scalability for Chatbot Integration: Mistral’s capabilities in text generation make it a strong candidate for powering a chatbot that responds to user queries about alternative parts. Its ability to handle conversational tasks and process structured prompts supports real-time applications, as outlined in the notebook’s integration goal.

## In summary, the Mistral model is chosen for its balance of performance, efficiency, and compatibility with the Hugging Face ecosystem, making it well-suited for both the similarity-matching task and its potential integration into a chatbot for real-time part recommendations.